File: examples/c/tiered.c
/*
 * turbolite example — C (S3 tiered)
 *
 * A sensor data logger that writes readings to S3 tiered SQLite.
 * Pages are compressed locally and synced to S3 for durability.
 *
 * Build & run:
 *   make lib-bundled header
 *   TURBOLITE_BUCKET=my-bucket make example-c-tiered
 *
 * Required:
 *   TURBOLITE_BUCKET          S3 bucket name
 *
 * Optional:
 *   TURBOLITE_ENDPOINT_URL    S3 endpoint (Tigris, MinIO, etc.)
 *   TURBOLITE_REGION          AWS region
 *   AWS_ACCESS_KEY_ID         S3 credentials
 *   AWS_SECRET_ACCESS_KEY     S3 credentials
 */

#include "../../turbolite.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

static void die(const char *msg) {
    const char *err = turbolite_last_error();
    fprintf(stderr, "FATAL: %s: %s\n", msg, err ? err : "unknown");
    exit(1);
}

static char tmpdir_buf[256];

int main(void) {
    const char *bucket = getenv("TURBOLITE_BUCKET");
    if (!bucket) {
        fprintf(stderr, "FATAL: TURBOLITE_BUCKET is required\n");
        return 1;
    }

    /* Create local cache directory */
    snprintf(tmpdir_buf, sizeof(tmpdir_buf), "%s/turbolite-tiered-XXXXXX",
             getenv("TMPDIR") ? getenv("TMPDIR") : "/tmp");
    const char *cache_dir = mkdtemp(tmpdir_buf);
    if (!cache_dir) { perror("mkdtemp"); return 1; }

    printf("turbolite %s — sensor logger (S3 tiered)\n", turbolite_version());
    printf("Bucket: %s\n", bucket);
    printf("Cache dir: %s\n\n", cache_dir);

    /* Register tiered VFS and open database */
    const char *endpoint = getenv("TURBOLITE_ENDPOINT_URL");
    const char *region = getenv("TURBOLITE_REGION");
    if (turbolite_register_tiered("tiered", bucket, "sensors", cache_dir, endpoint, region) != 0)
        die("register tiered VFS");

    char db_path[512];
    snprintf(db_path, sizeof(db_path), "%s/sensors.db", cache_dir);

    void *db = turbolite_open(db_path, "tiered");
    if (!db) die("open database");

    turbolite_exec(db,
        "CREATE TABLE IF NOT EXISTS readings ("
        "  id INTEGER PRIMARY KEY AUTOINCREMENT,"
        "  sensor TEXT NOT NULL,"
        "  value REAL NOT NULL,"
        "  ts INTEGER NOT NULL"
        ")");

    /* Simulate logging sensor readings */
    srand((unsigned)time(NULL));
    for (int i = 0; i < 10; i++) {
        char sql[256];
        double temp = 20.0 + (rand() % 100) / 10.0;
        snprintf(sql, sizeof(sql),
            "INSERT INTO readings (sensor, value, ts) VALUES ('temp-01', %.1f, %ld)",
            temp, (long)time(NULL) + i);

        if (turbolite_exec(db, sql) != 0)
            die("insert reading");
    }
    printf("Logged 10 readings.\n");

    /* Query summary */
    char *json = turbolite_query_json(db,
        "SELECT sensor, COUNT(*) as count, "
        "ROUND(AVG(value), 1) as avg_temp, "
        "ROUND(MIN(value), 1) as min_temp, "
        "ROUND(MAX(value), 1) as max_temp "
        "FROM readings GROUP BY sensor");
    if (!json) die("query");

    printf("Summary: %s\n", json);
    turbolite_free_string(json);

    turbolite_close(db);
    return 0;
}

File: examples/rust/src/tiered.rs
//! turbolite example — Rust S3 tiered (native API)
//!
//! A small CLI that stores and queries books in an S3 tiered SQLite database.
//! Pages are compressed locally and synced to S3 for durability.
//!
//! Run:
//!   TURBOLITE_BUCKET=my-bucket cargo run --bin tiered
//!
//! Required:
//!   TURBOLITE_BUCKET          S3 bucket name
//!
//! Optional:
//!   TURBOLITE_ENDPOINT_URL    S3 endpoint (Tigris, MinIO, etc.)
//!   TURBOLITE_REGION          AWS region
//!   AWS_ACCESS_KEY_ID         S3 credentials
//!   AWS_SECRET_ACCESS_KEY     S3 credentials

use rusqlite::{params, Connection, OpenFlags};
use turbolite::tiered::{register, TieredConfig, TieredVfs};
use tempfile::TempDir;

fn main() {
    let bucket = std::env::var("TURBOLITE_BUCKET")
        .expect("TURBOLITE_BUCKET is required");

    let cache_dir = TempDir::new().expect("create temp dir");

    let config = TieredConfig {
        bucket,
        prefix: "books".to_string(),
        cache_dir: cache_dir.path().to_path_buf(),
        endpoint_url: std::env::var("TURBOLITE_ENDPOINT_URL").ok(),
        region: std::env::var("TURBOLITE_REGION").ok(),
        ..Default::default()
    };

    let vfs = TieredVfs::new(config).expect("create tiered VFS");
    register("tiered", vfs).expect("register tiered VFS");

    let flags = OpenFlags::SQLITE_OPEN_READ_WRITE | OpenFlags::SQLITE_OPEN_CREATE;
    let db_path = cache_dir.path().join("books.db");
    let conn =
        Connection::open_with_flags_and_vfs(&db_path, flags, "tiered").expect("open database");

    conn.execute_batch(
        "CREATE TABLE IF NOT EXISTS books (id INTEGER PRIMARY KEY, title TEXT NOT NULL, year INTEGER);
         INSERT INTO books VALUES (1, 'Dune', 1965);
         INSERT INTO books VALUES (2, 'Neuromancer', 1984);
         INSERT INTO books VALUES (3, 'Snow Crash', 1992);",
    )
    .expect("create and insert");

    let mut stmt = conn
        .prepare("SELECT id, title, year FROM books ORDER BY year")
        .expect("prepare");

    let books: Vec<(i64, String, i64)> = stmt
        .query_map(params![], |row| Ok((row.get(0)?, row.get(1)?, row.get(2)?)))
        .expect("query")
        .collect::<Result<_, _>>()
        .expect("collect");

    for (id, title, year) in &books {
        println!("  {id}. {title} ({year})");
    }

    let recent: Vec<String> = conn
        .prepare("SELECT title FROM books WHERE year > ?1")
        .expect("prepare")
        .query_map(params![1980], |row| row.get(0))
        .expect("query")
        .collect::<Result<_, _>>()
        .expect("collect");

    println!("\nBooks after 1980: {recent:?}");
}
